### ISSUE TITLE
reason command doesn't infer types on individuals

### ISSUE THREAD

**Issue by dosumis** [ontodev/robot#273]
I've been testing out robot on the VFB pipeline. This requires instantiation of inferred SubClassOf and Type axioms to named classes, followed by redundancy stripping for both.  'reason' appears to work fine for SubClassOf axioms but not for Type axioms.  Can this be fixed?

**Comment by cmungall** [comment]
Straightforward fix, I also added equivalence axioms.

Overall, I think it makes sense to do all possible axiom types by default. This would mean things like domain/range on subProperties - but I think this would be really useful.

We need to check that this doesn't violate any existing expectations. If it does we can always make subClassOf default, and the others optional - but seems mildly asymmetrical.

I'd also like the ability to do indirect as well as direct. Obviously this would *not* be the default.

**Comment by dosumis** [comment]
> I think it makes sense to do all possible axiom types by default

Agreed.  Can't see any reason not to.

**Comment by balhoff** [comment]
Some of them can slow things down quite a bit—it should definitely be configurable. If you add `InferredPropertyAssertionGenerator`, ELK will throw an exception.

**Comment by dosumis** [comment]
> And then there is subClassOf with materialized SVFs, which is currently handled separately by materialize which is a bit awkward

Happy to keep that as a separate command.  Would be useful if 'materialize' generated Type assertions too.

**Comment by matentzn** [comment]
Hey all, 

I have implemented this functionality (materialisation of various axiom types) a couple of times in [other tools](https://www.sciencedirect.com/science/article/pii/S1570826817300367), and there are a bunch of considerations to be made here.  

1.  ELK is a super-efficient classifier, but it lacks a lot of support for other axiom types. In particular, Version 0.4.3, the one ROBOT relies on, does not have a "isEntailedBy" method implementation, which is required for most InferredAxiom generators. However, there seems to be something in the latest version of ELK (0.5) that looks a bit more [promising](https://github.com/liveontologies/elk-reasoner/blob/a91843c08ea191a2863d6e5af63327936a28f9db/elk-owlapi-parent/elk-owlapi4/src/main/java/org/semanticweb/elk/owlapi/ElkReasoner.java#L1004), but I have not checked it yet. The same goes for getSubObjectProperties(): nothing in Elk 0.4.3, but something in Elk 0.5. There is, however, virtually no support for the many other more fancy axiom types: Inverses (of course), Property Domains (you could cheat this by using isEntailedBy), disjoint classes (!), object property assertions. I dont think there is even a ObjectPropertyDomainAxiomGenerator anyways (of course, EVERY missing axiom type can be extracted using isEntailedBy(), but it will require a lot of work to make sure that works properly, and maybe be very slow). 
2. The size of the transitive closure (indirect inferences) of the class graph, but more importantly of the object property relation (all inferred a R b) is potentially exponential, and often enormous; it only rarely makes sense to assert it. 
3. Even if you use HermiT (which can barely classify some of your larger ontologies overnight, and often not without 200GB+ memory), there are certain things you should be aware of: there is currently no efficient algorithm to compute the transitive reduct (i.e. most specific, direct) set of class disjointnesses, at least last I looked. It can, and will, take ages, many more times than to classify, because we have currently no good way to reduce the search space; we have to test almost all pairs of classes to be sure we catch all implied disjointnesses. Given that HermiT is already very slow to classify, trying to compute inferred disjoint classes, let alone the transitive reduct, will be feasible only for simple ontologies (small and inexpressive), at least in my experience. 
4. If you are going to use ROBOT reason for quality control (i.e. diff inferred class hierarchy over version etc), you should think of developing a formal description of the entailment sets you are using; I can help you do that. I have described this [elsewhere](https://github.com/OBOFoundry/OBOFoundry.github.io/issues/542) in more detail; There is no such thing as an "inferred ontology", nor is there a canonical representation of the class graph.

I am happy to chip in on this, but we should first get clear of what the role of the "reason" command for the ontology development life cycle will be. IMHO, deploying "inferred ontologies" is really misleading: It suggests you can somehow get inferences from it without using the reasoner. This is only barely true in one extreme case: the transitive closure of absolutely all axioms types are asserted. But not really even there. Since we usually simply assert the transitive reduct of the class graph, maybe some equivalences (i.e. direct subsumption), this is certainly not given at all. However, I do see *a lot* of benefit to use materialised ontologies for quality control, and for adding inferred subclasses as a means to browse the ontology without having to classify (for OWL API based tools like OLS and Protege). 

My suggestion is to keep the three key entailment sets that are really important for quality control:

* Inferred subsumption
* Inferred equivalence
* Inferred typing

And define entailment sets for each of them (I did so crudely [here](https://www.sciencedirect.com/science/article/pii/S1570826817300367)). Allow perhaps a choice between closure (indirect) and reduct (direct). Allow to exclude subsumptions that are unrelated to the signature of the root ontology would be a good thing too.

**Comment by cmungall** [comment]
Thanks @matentzn, +1 to all your points. I'm convinced assert-all should not be the default. I like your default entailment set.

Re: indirect. The main use case here is not QC but building a lookup table for dumb apps, solr indexing etc. It usually makes sense to serialize these as simple compact TSVs.

**Comment by cmungall** [comment]
all excellent points @matentzn 

Hmm, I was sure I wrote a detailed response. But yes I like the idea of entailment sets and your proposed default is a good one

> Allow to exclude subsumptions that are unrelated to the signature of the root ontology would be a good thing too

Yes! I currently have to resort to all manner of hacks to filter out inferred subClassOf between classes in an import being asserted in the importing ontology.

**Comment by balhoff** [comment]
Just wanted to vote for optionally requesting additional axiom generators on the command line. I have a need for asserting inferred DisjointWith, although it will be slow.

**Comment by rctauber** [comment]
Hi all - I just pushed a some changes to a new branch to add an `--axiom-generators` option to `reason`: https://github.com/rctauber/robot/tree/axiom-generator

If anybody has time to test it out and give feedback, that would be great. Information on the option below.

---
## Axiom Generators

By default, the `reason` operation will only assert inferred subclass axioms. This can be configured with the `--axiom-generators` option. OWLAPI provides the following inferred axiom generators:

  * Class axiom generators
    * `SubClass`
    * `EquivalentClass`
    * `DisjointClasses`
  * Data property axiom generators
    * `DataPropertyCharacteristic`
    * `EquivalentDataProperties`
    * `SubDataProperty`
  * Individual axiom generators
    * `ClassAssertion`
    * `PropertyAssertion`
  * Object property axiom generators
    * `EquivalentObjectProperties`
    * `InverseObjectProperties`
    * `ObjectPropertyCharacteristic`
    * `SubObjectProperty`
  
One or more of these axiom generators can be passed into the option (separated by spaces). For example, to generate both subclass and disjoint class axioms:

```
robot reason --input unreasoned.owl
  --axiom-generators "SubClass DisjointClasses"
  --output reasoned.owl
```

**Comment by cmungall** [comment]
Awesome, can we have an orthogonal option to also create indirect? (which
may only apply to SubClass, I forget if the Equivalence generator makes the
complete clique)

>

**Comment by balhoff** [comment]
> Awesome, can we have an orthogonal option to also create indirect? (which
may only apply to SubClass

And ClassAssertion.

**Comment by cmungall** [comment]
Ah yes of course! And ObjectPropertyAssertion too (won't work for Elk but
will for Arachne)

**Comment by rctauber** [comment]
Cool. I added an `--include-indirect <boolean>` option. Here's the burb from the docs:
> The axioms that are generated for subclasses, class assertions, and sub-object properties can be either direct or indirect. By default, the operation will only generate direct axioms (`--include-indirect false`). For example, to generate all direct and indirect class assertions:
```
robot reason --input unreasoned.owl
  --axiom-generators ClassAssertion
  --include-indirect true
  --output reasoned.owl
```

So this will add indirect for: `SubClass`, `ClassAssertion`, and `SubObjectProperty`. Is there anything else that should be included?

**Comment by rctauber** [comment]
I don't see Arachne support anywhere in the OWLAPI, would we still be able to add the object property assertions?

**Comment by balhoff** [comment]
We can add Arachne at some point in the future. We want to have `ObjectPropertyAssertion`, just since HermiT and others support it. But I think you already meant to include it where you wrote `PropertyAssertion` under individual axioms.

**Comment by rctauber** [comment]
Yeah - we have the `PropertyAssertion` feature, but I can't do anything with the `--include-indirect <boolean>` because I can't get indirect object property assertions, as far as I know.

**Comment by balhoff** [comment]
I'm not sure @cmungall is right about that—as far as I know you get all object property assertions, whether or not one property is a subproperty of another.

**Comment by cmungall** [comment]
I had just assumed that there was direct and indirect options with direct
not including transitive closure. But that's just how I would implement it
myself but I haven't looked at this in the Reasoner API for a while. I
vaguely recall writing mw own filter to get rid of indirects

>

**Comment by matentzn** [comment]
With object property assertions, there is no 100% clean way to define what direct and what indirect would mean I think. You have so many axes of indirection, like property hierarchy, transitivity, and so on; you have cyclicity; At least instinctively, I would think that a transitive reduct (or some notion of directness) only makes sense for directed graphs. I agree with @balhoff I think.

**Comment by cmungall** [comment]
> The only problem I see with the "indirect" option is that it can add a lot of redundant axioms. We could throw reduce in the pipeline, or recommend that people reduce after using --include-indirect true?

indirect = redundant and non-redundant, the intent is to have these for query purposes (where we want to give dumb apps a pre-reasoned graph they can use as a lookup table)

**Comment by jamesaoverton** [comment]
This should be addressed by #439. Please try it out, and open a new issue if you run into any problems.
